pkg/tools/logging/query.go

```diff
@@ -0,0 +1,280 @@
+	"google.golang.org/protobuf/encoding/protojson"
+)
+
+type LogQueryRequest struct {
+	Query     string     `json:"query"`
+	ProjectID string     `json:"project_id"`
+	TimeRange *TimeRange `json:"time_range,omitempty"`
+	Since     string     `json:"since,omitempty"`
+	Limit     int        `json:"limit,omitempty"`
+	Format    string     `json:"format,omitempty"`
+}
+
+type TimeRange struct {
+	StartTime time.Time `json:"start_time"`
+	EndTime   time.Time `json:"end_time"`
+}
+
+const (
+	defaultLimit = 10
+	maxLimit     = 100
+)
+
+func installQueryLogsTool(s *server.MCPServer, conf *config.Config) {
+	queryLogsTool := mcp.NewTool("query_logs",
+		mcp.WithDescription("Query Google Cloud Platform logs using Logging Query Language (LQL). Before using this tool, it's **strongly** recommended to call the 'get_log_schema' tool to get information about supported log types and their schemas. Logs are returned in ascending order, based on the timestamp (i.e. oldest first)."),
+		mcp.WithReadOnlyHintAnnotation(true),
+		mcp.WithString("project_id", mcp.Description("GCP project ID to query logs from. Required."), mcp.Required()),
+		mcp.WithString("query", mcp.Description("LQL query string to filter and retrieve log entries. Don't specify time ranges in this filter. Use 'time_range' instead. Consider using https://cloud.google.com/logging/docs/view/logging-query-language to learn more about LQL syntax.")),
+		mcp.WithObject("time_range", mcp.Description("Time range for log query. If empty, no restrictions are applied."),
+			mcp.Properties(map[string]any{
+				"start_time": map[string]any{
+					"type":        "string",
+					"description": "Start time for log query (RFC3339 format)",
+				},
+				"end_time": map[string]any{
+					"type":        "string",
+					"description": "End time for log query (RFC3339 format)",
+				},
+			}),
+		),
+		mcp.WithString("since", mcp.Description("Only return logs newer than a relative duration like 5s, 2m, or 3h. The only supported units are seconds ('s'), minutes ('m'), and hours ('h').")),
+		mcp.WithNumber("limit", mcp.Description(fmt.Sprintf("Maximum number of log entries to return. Cannot be greater than %d. Consider multiple calls if needed. Defaults to %d.", maxLimit, defaultLimit))),
+		mcp.WithString("format", mcp.Description("Go template string to format each log entry. If empty, the full JSON representation is returned. Note that empty fields are not included in the response. Example: '{{.timestamp}} [{{.severity}}] {{.textPayload}}'. It's strongly recommended to use a template to minimize the size of the response and only include the fields you need. Use the get_schema tool before this tool to get information about supported log types and their schemas.")),
+	)
+
+	t := NewQueryLogsTool(conf)
+	s.AddTool(queryLogsTool, mcp.NewTypedToolHandler(t.QueryLogs))
+}
+
+type QueryLogsTool struct {
+	conf *config.Config
+}
+
+func NewQueryLogsTool(conf *config.Config) *QueryLogsTool {
```

Comment:
I don't think this needs to be public / exported.

Reply:
Done.